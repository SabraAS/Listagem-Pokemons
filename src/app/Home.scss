.home {
  padding: 64px;

  &__title {
    font-size: 32px;
    font-weight: 700;
    margin-bottom: 24px;
  }

  &__content {
    display: flex;
    flex-direction: row;
    align-items: flex-start;
    justify-content: space-between;
    width: 100%;
    gap: 24px;

    @media (max-width: 768px) {
      gap: 0;
    }
  }

  &__list {
    display: flex;
    flex-direction: row;
    gap: 24px;
    align-items: flex-start;
    justify-content: flex-start;
    flex-wrap: wrap;
    max-width: calc(100% - 500px);
    width: 100%;

    @media (max-width: 768px) {
      flex-direction: column;
      align-items: center;
      justify-content: center;
      max-width: 100%;
    }
  }
}
