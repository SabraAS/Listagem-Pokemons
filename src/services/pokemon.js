import axios from 'axios';

const BASE_URL = 'https://pokeapi.co/api/v2';

export const getPokemonList = async (limit = 40) => {
  try {
    const response = await axios.get(`${BASE_URL}/pokemon?limit=${limit}`);
    const pokemonList = response.data.results;

    return await Promise.all(
      pokemonList.map(async (pokemon) => {
        const pokemonData = await axios.get(pokemon.url);
        const description = await getPokemonCharacteristic(pokemonData.data.id);
        return {
          id: pokemonData.data.id,
          name: pokemonData.data.name,
          image: pokemonData.data.sprites.front_default,
          abilities: pokemonData.data.abilities,
          types: pokemonData.data.types,
          characteristic: description,
        };
      }),
    );
  } catch (error) {
    console.error('Erro ao buscar lista de Pokémon:', error);
    throw error;
  }
};

export const getPokemonCharacteristic = async (id) => {
  try {
    const response = await fetch(
      `https://pokeapi.co/api/v2/characteristic/${id}`,
    );
    const data = await response.json();
    const englishDescription = data.descriptions.find(
      (desc) => desc.language.name === 'en',
    );
    return englishDescription?.description || '';
  } catch (e) {
    console.error('Erro ao buscar característica do Pokémon:', e);
    return '';
  }
};
