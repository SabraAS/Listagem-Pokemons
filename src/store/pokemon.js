import { create } from 'zustand';

export const usePokemonStore = create((set) => ({
  pokemons: [],
  setPokemons: (pokemons) => set({ pokemons }),
  addPokemon: (pokemon) =>
    set((state) => ({ pokemons: [...state.pokemons, pokemon] })),
  removePokemon: (pokemonId) =>
    set((state) => ({
      pokemons: state.pokemons.filter((pokemon) => pokemon.id !== pokemonId),
    })),
  clearTeam: () => set({ pokemons: [] }),
}));
