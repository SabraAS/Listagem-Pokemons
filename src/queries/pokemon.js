import { useQuery } from '@tanstack/react-query';

import { getPokemonList } from '@/services/pokemon';

export const POKEMON_KEYS = {
  all: ['pokemons'],
  list: (limit) => [...POKEMON_KEYS.all, 'list', limit],
};

export const usePokemons = (limit = 20) => {
  return useQuery({
    queryKey: POKEMON_KEYS.list(limit),
    queryFn: () => getPokemonList(limit),
    retry: false,
    refetchOnWindowFocus: false,
    select: (response) => response,
  });
};
