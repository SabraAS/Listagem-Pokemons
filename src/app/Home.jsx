import './Home.scss';

import CartSidebar from '@/components/CartSidebar';
import PokemonCard from '@/components/PokemonCard';
import { usePokemons } from '@/queries/pokemon';
import { usePokemonStore } from '@/store/pokemon';

const Home = () => {
  const { data } = usePokemons(30);
  const { pokemons, addPokemon, removePokemon, clearTeam } = usePokemonStore();

  return (
    <div className="home">
      <h1 className="home__title">Pok√©mons</h1>
      <div className="home__list">
        {data &&
          data?.map((pokemon) => (
            <PokemonCard
              abilities={pokemon.abilities}
              characteristic={pokemon.characteristic}
              disabled={Boolean(pokemons.find((p) => p.id === pokemon.id))}
              image={pokemon.image}
              key={pokemon.id}
              name={pokemon.name}
              onClick={() => {
                addPokemon(pokemon);
              }}
              types={pokemon.types}
            />
          ))}
      </div>
      <CartSidebar
        clearTeam={clearTeam}
        pokemons={pokemons}
        removePokemon={removePokemon}
      />
    </div>
  );
};

export default Home;
