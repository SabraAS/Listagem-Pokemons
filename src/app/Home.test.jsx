import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';

import Home from './Home';

import { usePokemons } from '@/queries/pokemon';
import { usePokemonStore } from '@/store/pokemon';
import { mockPokemons } from '@/test/mocks/pokemon';

// Mock dos módulos
vi.mock('@/queries/pokemon', () => ({
  usePokemons: vi.fn(),
}));

vi.mock('@/store/pokemon', () => ({
  usePokemonStore: vi.fn(),
}));

describe('Home Component', () => {
  const queryClient = new QueryClient();
  const mockAddPokemon = vi.fn();
  const mockRemovePokemon = vi.fn();
  const mockClearTeam = vi.fn();

  beforeEach(() => {
    // Reset all mocks before each test
    vi.clearAllMocks();

    // Setup default mock implementations
    usePokemons.mockReturnValue({
      data: mockPokemons,
      isLoading: false,
    });

    usePokemonStore.mockReturnValue({
      pokemons: [],
      addPokemon: mockAddPokemon,
      removePokemon: mockRemovePokemon,
      clearTeam: mockClearTeam,
    });
  });

  describe('Rendering', () => {
    it('should render the title correctly', () => {
      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      expect(screen.getByText('Pokémons')).toBeInTheDocument();
    });

    it('should render pokemon cards when data is loaded', () => {
      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      expect(screen.getByText('bulbasaur')).toBeInTheDocument();
      expect(screen.getByText('charmander')).toBeInTheDocument();
    });

    it('should render loading message when data is not loaded', () => {
      usePokemons.mockReturnValue({
        data: [],
        isLoading: true,
      });

      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      expect(screen.getByText('Carregando...')).toBeInTheDocument();
    });
  });

  describe('Interactions', () => {
    it('should call addPokemon when clicking on a pokemon card', () => {
      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const firstPokemonCard = screen.getByTestId('add-pokemon-bulbasaur');
      fireEvent.click(firstPokemonCard);

      expect(mockAddPokemon).toHaveBeenCalledWith(mockPokemons[0]);
    });

    it('should disable pokemon card if pokemon is already in store', () => {
      usePokemonStore.mockReturnValue({
        pokemons: [mockPokemons[0]],
        addPokemon: mockAddPokemon,
        removePokemon: mockRemovePokemon,
        clearTeam: mockClearTeam,
      });

      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const firstPokemonCard = screen.getByTestId('add-pokemon-bulbasaur');
      expect(firstPokemonCard).toBeDisabled();
    });
  });

  describe('Accessibility', () => {
    it('should have no accessibility violations on home page', async () => {
      const { container } = render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should have no accessibility violations with disabled pokemon cards', async () => {
      usePokemonStore.mockReturnValue({
        pokemons: [mockPokemons[0]], // First pokemon already selected
        addPokemon: mockAddPokemon,
        removePokemon: mockRemovePokemon,
        clearTeam: mockClearTeam,
      });

      const { container } = render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should have proper keyboard navigation', () => {
      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const firstPokemonButton = document.querySelector(
        '[data-testid="add-pokemon-bulbasaur"]',
      );
      expect(firstPokemonButton).not.toBeNull();
      expect(firstPokemonButton.tagName).toBe('BUTTON');
      expect(firstPokemonButton.getAttribute('tabindex')).not.toBe('-1');
    });

    it('should have proper ARIA labels and roles', () => {
      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const buttons = document.querySelectorAll('button');
      buttons.forEach((button) => {
        expect(button.getAttribute('role')).toBeNull(); // Buttons should not need explicit role
        expect(button.textContent.trim().length).toBeGreaterThan(0); // Should have text content
      });
    });

    it('should have proper alt text for images', () => {
      render(
        <QueryClientProvider client={queryClient}>
          <Home />
        </QueryClientProvider>,
      );

      const images = document.querySelectorAll('img');
      images.forEach((img) => {
        expect(img.getAttribute('alt')).not.toBeNull();
        expect(img.getAttribute('alt').trim().length).toBeGreaterThan(0);
      });
    });
  });
});
