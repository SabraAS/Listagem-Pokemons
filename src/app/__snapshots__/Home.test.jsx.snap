// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should match snapshot 1`] = `
<div>
  <div
    class="home"
  >
    <h1
      class="home__title"
    >
      Pokémons
    </h1>
    <div
      class="home__content"
    >
      <ul
        class="home__list"
      >
        <li>
          <div
            class="pokemon-card"
            data-testid="pokemon-card"
          >
            <img
              alt="Imagem do bulbasaur"
              class="pokemon-card__image"
              data-testid="pokemon-card-image"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
            />
            <button
              aria-label="Adicionar bulbasaur à equipe"
              class="pokemon-card__button"
              data-testid="pokemon-card-button"
            >
              Adicionar à equipe
            </button>
            <h2
              class="pokemon-card__name"
              data-testid="pokemon-card-name"
            >
              bulbasaur
            </h2>
            <div
              class="pokemon-card__info"
            >
              <p
                class="pokemon-card__text"
                data-testid="pokemon-card-characteristic"
              >
                Característica: 
                A strange seed was planted on its back at birth. The plant sprouts and grows with this Pokémon.
              </p>
              <p
                class="pokemon-card__text"
                data-testid="pokemon-card-abilities"
              >
                Habilidades: 
                overgrow, chlorophyll
              </p>
              <p
                class="pokemon-card__text"
                data-testid="pokemon-card-types"
              >
                Tipos: 
                grass, poison
              </p>
            </div>
          </div>
        </li>
        <li>
          <div
            class="pokemon-card"
            data-testid="pokemon-card"
          >
            <img
              alt="Imagem do charmander"
              class="pokemon-card__image"
              data-testid="pokemon-card-image"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/4.svg"
            />
            <button
              aria-label="Adicionar charmander à equipe"
              class="pokemon-card__button"
              data-testid="pokemon-card-button"
            >
              Adicionar à equipe
            </button>
            <h2
              class="pokemon-card__name"
              data-testid="pokemon-card-name"
            >
              charmander
            </h2>
            <div
              class="pokemon-card__info"
            >
              <p
                class="pokemon-card__text"
                data-testid="pokemon-card-characteristic"
              >
                Característica: 
                Obviously prefers hot places. When it rains, steam is said to spout from the tip of its tail.
              </p>
              <p
                class="pokemon-card__text"
                data-testid="pokemon-card-abilities"
              >
                Habilidades: 
                blaze, solar-power
              </p>
              <p
                class="pokemon-card__text"
                data-testid="pokemon-card-types"
              >
                Tipos: 
                fire
              </p>
            </div>
          </div>
        </li>
        <li>
          <div
            class="home__loader"
            data-testid="loader"
          />
        </li>
      </ul>
      <aside
        class="cart-sidebar"
        data-testid="cart-sidebar"
      >
        <img
          alt="Imagem de fundo da sidebar"
          class="cart-sidebar__background"
          data-testid="cart-sidebar-background"
          src="/src/assets/cover-pokemon.png"
        />
        <h2
          class="cart-sidebar__title"
          data-testid="cart-sidebar-title"
        >
          Sua equipe
        </h2>
        <div
          class="cart-sidebar__list"
          data-testid="cart-sidebar-list"
        >
          <p
            data-testid="cart-sidebar-empty-message"
          >
            Nenhum Pokémon adicionado
          </p>
        </div>
        <div
          class="cart-sidebar__footer"
          data-testid="cart-sidebar-footer"
        >
          <button
            class="cart-sidebar__footer-button"
            data-testid="cart-sidebar-footer-button"
            disabled=""
          >
            Confirmar Equipe
          </button>
        </div>
      </aside>
    </div>
  </div>
</div>
`;
